src/rust/plugin-registry/src/server/service.rs

```diff
@@ -129,32 +149,76 @@ impl PluginRegistryApi for PluginRegistry {
             tenant_id,
-            plugin_artifact, // This could be huge, so don't clone it!
             display_name,
             plugin_type,
-        } = request;
-
-        self.ensure_artifact_size_limit(&plugin_artifact)?;
-
-        let plugin_id = generate_plugin_id(&tenant_id, plugin_artifact.as_slice());
+        } = match request.next().await {
+            Some(CreatePluginRequest::Metadata(m)) => m,
+            _ => {
+                return Err(Self::Error::StreamInputError(
+                    "Expected request 0 to be Metadata",
+                ));
+            }
+        };
 
+        let plugin_id = generate_plugin_id();
         let s3_key = generate_artifact_s3_key(plugin_type, &tenant_id, &plugin_id);
 
+        // Convert the incoming request stream of CreatePluginRequest::Chunk
+        // into a stream of Bytes to be sent to Rusoto S3.put_object.
+        let body_stream = request.then(|result| async {
+            match result {
+                CreatePluginRequest::Chunk(c) => Ok(Bytes::from(c)),
+                _ => Err(io_input_error("Expected request 1..N to be Chunk")),
+            }
+        });
+
+        // While Rusoto reads the stream, keep track of total legth of chunks.
+        // Bail out if we exceed the limit specified in the config.
+        let limit_bytes = self.config.artifact_size_limit_mb.clone() * 1024 * 1024;
+        let stream_length = Arc::new(std::sync::atomic::AtomicUsize::new(0));
+        // Get a handle for the Stream to move
+        let stream_length_handle = stream_length.clone();
+        let body_stream = body_stream.then(move |result| {
+            accumulate_stream_size_to_limit(stream_length_handle.clone(), limit_bytes, result)
+        });
+
+        // Send the Stream into Rusoto, which does the actual awaiting of each
+        // future in the Stream.
         self.s3
             .put_object(PutObjectRequest {
-                content_length: Some(plugin_artifact.len() as i64),
-                body: Some(StreamingBody::from(plugin_artifact)),
+                body: Some(StreamingBody::new(body_stream)),
                 bucket: self.config.bucket_name.clone(),
                 key: s3_key.clone(),
                 expected_bucket_owner: Some(self.config.bucket_aws_account_id.clone()),
                 ..Default::default()
             })
             .await?;
 
+        // Emit some benchmark info
+        {
+            let stream_length: usize = Arc::try_unwrap(stream_length)
+                .expect("Stream has been exhausted by this point")
+                .into_inner();
+            let total_duration = std::time::SystemTime::now()
+                .duration_since(start_time)
+                .unwrap_or_default();
+
+            tracing::info!(
+                message = "CreatePlugin benchmark",
+                display_name = ?display_name,
+                duration_millis = ?total_duration.as_millis(),
+                size_mb = stream_length / (1024 * 1024)
```

Comment:
For a debug log I think it'd make more sense to just put `stream_length` in there directly rather than divide

Reply:
done